working/augmentation-libraries/feature-specification.md

```diff
@@ -622,311 +626,97 @@ It is a compile-time error if:
-has the name `values`, `index`, `hashCode`, or `==`.
-
-*It has always been an error for an enum declaration to declare a member
-named `index`, `hashCode`, `==`, or `values`, and this rule just clarifies
-that this error is applicable for augmenting declarations as well.*
-
-Enum values can _only_ be augmented by enum values, and the implicit getter
-introduced by them is not augmentable. The only thing you are allowed to do
-when augmenting an enum value is add metadata annotations or doc comments.
+### Augmenting variables
 
-When augmenting an enum value, no constructor invocation should be provided.
-The original value is always used, and the explicit constructor invocation (if
-present) should not be copied.
+A class-like augmentation can add *new* variables (i.e. instance or static
+fields) to the type being augmented:
 
-New enum values may be defined in an augmenting enum, and they will be appended
-to the current values of the declaration in augmentation application order.
-
-Augmenting an existing enum value never changes the order in which it appears in
-`values`.
+```dart
+class C {}
 
-For example:
+augment class C {
+  int x = 3;
 
+  static int y = 4;
+}
 ```
-// main.dart
-part 'a.dart';
-part 'c.dart';
 
-enum A {
-  first,
-  second.custom(1);
+Variable declarations themselves can't be directly augmented, with the exception
+of abstract fields.
 
-  final int b;
+*A variable declaration implicitly has code for the synthesized getter and
+setter that access and modify the underlying backing storage. Since we don't
+allow augmentations to replace code, that implies that augmentations can't
+change variables. So we don't allow them to be augmented.*
 
-  const A() : b = 0;
+#### Abstract fields
 
-  const A.custom(this.b);
-}
-}
+Dart supports `abstract` field declarations. They are syntactic sugar for
+declaring an abstract getter with an optional abstract setter if the variable is
+non-final. They don't actually declare a variable with any backing storage.
 
-// a.dart
-part of 'main.dart';
-part 'b.dart';
+Because abstract variables are effectively abstract getter and setter
+declarations, they can be augmented and used in augmentations just like function
+declarations:
 
-augment enum A {
-  third;
+*Examples:*
 
-  /// Some doc comment
-  augment first; // This is still `first` in values.
+```dart
+class C {
+  // Augment an abstract variable with a getter:
+  abstract final int a;
+  augment int get a => 1; // OK.
 
-  @someAnnotation
-  augment second; // Don't repeat the argument list, original is used.
-}
+  // Augment an abstract variable with a getter and setter:
+  abstract int b;
+  augment int get b => 1; // OK.
+  augment set b(int value) {} // OK.
 
-// b.dart
-part of 'a.dart';
+  // Augment a getter with an abstract variable:
+  int get c;
 
-augment enum A {
-  fourth;
+  @someMetadata
+  augment abstract final int c; // (Not very useful, but valid.)
 }
+```
```

Comment:
Might need to say how type inheritance work if the getter and setter have different types, and if the declaration has more than one variable:
```dart
int get x;
set x(num _);
abstract String y;
@markOfShame
augment abstract x, y;
```
(Maybe it's not a problem, we use the same inheritance rules as we would for an override in a subclass ... which don't handle this either;
```dart
abstract class C {
  int get x;
  set x(num _);
  abstract String y;
}

abstract class D implements C {
  abstract var x, y;
}
```
makes `x` an error bacause it cannot inherit a combined signature.)

Reply:
Good catch. Added a section under type annotation inheritance to talk about this.